### ISSUE TITLE
[1.0.0.26][Carol] Walljump does not function on some Surfaces if in a narrow space

### ISSUE THREAD

**Issue by Deckardsmegson** [whitelilydragon/ShangMuArchitect#206]
https://cdn.discordapp.com/attachments/739994967186997329/815389411234676756/2021-02-28_02-05-47.mp4
https://cdn.discordapp.com/attachments/739994967186997329/815389188386979840/2021-02-28_01-49-12.mp4

Carol can not wall-jump on the shown tiles if the space is too narrow

**Comment by whitelilydragon** [comment]
I am labeling this as #Won't Fix, and I'll explain why: It is a problem both Freedom Planet and video games like Sonic face at large, and I'm not capable of providing a meaningful answer to that myself.

Watch this footage below:

https://user-images.githubusercontent.com/23007238/109406912-cf74a100-794a-11eb-851d-625bb2a7f949.mov

You'll notice that this is 32x32 pixel divot in the ground that Milla is standing in, and all of her active surface surfaces are in an absolute frenzy to push her away from the surrounding walls. (This is how Damizean's collision model works! It pushes players away from surfaces at a 0 pixel-level precision above their surfaces' edges, no matter how curved.)

These sensors are all placed rather far apart, because of how large the character sprites are. Each major surface sensor in Freedom Planet is at least 16 pixels away from the center of the character's sprite, whereas Shang Mu Architect has each wall surface sensor placed 15 pixels away. (This is why standing in such a similarly tight space won't immediately cause problems in SMA, like it does in FP1.)

If you created a divot any thinner than 32 pixels wide in FP1, (which is the size of the editor's visual grid!) you'd either risk clipping issues, or crushing/killing the player, or much worse than this. This is why, in most cases, the game's levels are designed around this limitation, instead of fixed, because it's just not a practical use of time, and may create performance problems if the logic to prioritize surfaces, instead of push away from all of them at once, becomes reasonably complex.

This problem is *only mildly* better in Shang Mu Architect. The game's sensors clip precisely against the edges of surfaces, instead of pushing the player away, which comes with different problems, but it does positively affect how the game interacts with thinner nooks and crannies.

However, this also means that how thinner nooks are handled isn't much better. Yes, while you can jump through a 24 pixel wide tunnel without issues, both left and right wall sensors are already pushed to their absolute limit. And neither one can be made contextually-aware enough to choose which side, left or right, to prioritize without risking unexpected clipping issues on either side of the player.

I can't risk shrinking the distance of the wall sensors while Carol's wall jumping phase is active, either, because the second she jumps away from a wall facing to her right, for instance, she might just clip into the wall on her left and promptly die.

What I can say, at least, is that the code for Carol's wall jumping was a little crusty, so I updated its conditions to prioritize which wall to cling to based on your given direction, which does make 32 pixel wide corridors work for her now! But, it's also in conflict with the sensors as detailed, like so:

https://user-images.githubusercontent.com/23007238/109407248-38f5af00-794d-11eb-86d2-00853373b808.mov

I mitigated 32x32 sized surfaces, but I can't go any further than that *without potentially breaking the entire game*. This is not how Freedom Planet works, it is not how Sonic works, either; not without a radically different collision system that can somehow rectify the existence of pushing animations and extremely wide sprites. There's no reasonable way to support this without it looking visually clunky in Freedom Planet's context, even besides the fact that it'd create an endless pandora's box of bugs to fix.

Games like Super Mario Maker entirely cheats problems like this by very thin characters, and a locked grid-size you can't deviate from while placing or tiling objects or terrain. Shang Mu Architect has to offer an 8x8 grid *by policy of how Freedom Planet works*, so there's room for problems like this to be discovered by players.

If I forced everyone to work with a 32x32 grid, just to avoid showing this to the casual player, the game would suddenly undermine its own premise and accuracy goals, and be rendered the biggest waste of four years of my life.

At the very least, I placed warning message down for scenarios like this: Transparency can do a lot to keep players' expectations in line with the technical aspects of a game.

https://user-images.githubusercontent.com/23007238/109407782-ccc97a00-7951-11eb-8c66-6b9d40de1779.mov

tl;dr, this is a enigma with no easy answer, so consider making those wooden edge tiles into scenery decorations.
